surface/include/pcl/surface/impl/ball_pivoting.hpp

```diff
@@ -0,0 +1,629 @@
+   */
+  template<typename PointNT>
+  Eigen::Vector3f
+  get_normal_triangle (const typename pcl::PointCloud<PointNT>::ConstPtr &cloud, 
+                       const std::vector<uint32_t> &indexes)
+  {
+    if (indexes.size () != 3)
+    {
+      PCL_ERROR ("indexes.size () != 3 in get_normal_triangle");
+      return Eigen::Vector3f::Zero ();
+    }
+    const Eigen::Vector3f p0 = cloud->at (indexes.at (0)).getVector3fMap ();
+    return (cloud->at (indexes.at (1)).getVector3fMap () - p0)
+      .cross (cloud->at (indexes.at (2)).getVector3fMap () - p0)
+      .normalized ();
+  }
+  
+  /**
+   * reorder the vertices of triangle so the normal vector of triangle is consistent with normals of vertices
+   * @tparam PointNT
+   * @param cloud
+   * @param triangle
+   */
+  template<typename PointNT>
+  void
+  reorder (const typename pcl::PointCloud<PointNT>::ConstPtr &cloud, pcl::Vertices &triangle)
+  {
+    if (!is_normal_consistent (get_normal_triangle<PointNT> (cloud, triangle.vertices), 
+                               triangle.vertices, cloud))
+    {
+      // if the order (0,1,2) is not consistent, (0,2,1) should be consistent
+      std::swap (triangle.vertices.at (1), triangle.vertices.at (2));
+    }
+  }
+  
+  /**
+   * get the distance from point to plane
+   * @tparam PointNT
+   * @param plane
+   * @param point
+   * @return
+   */
+  template<typename PointNT>
+  float
+  get_distance_point_plane (const Eigen::Vector4f &plane, const PointNT &point)
+  {
+    // plane is (a,b,c,d), point is (x,y,z), then distance is ax+by+cz+d
+    return point.getVector3fMap ().dot (plane.segment (0, 3)) + plane (3);
+  }
+  
+  /**
+   * get the signed rotation angle from (point0-center) to (point1-center) on plane
+   * @param point0
+   * @param point1
+   * @param center
+   * @param plane the rotation is along the normal vector of plane
+   * @return
+   */
+  float
+  get_angle_rotation (const Eigen::Vector3f &point0, const Eigen::Vector3f &point1, 
+                      const Eigen::Vector3f &center, const Eigen::Vector4f &plane)
+  {
+    const Eigen::Vector3f vc0 = (point0 - center).normalized ();
+    const Eigen::Vector3f vc1 = (point1 - center).normalized ();
+  
+    const float sin_val = vc0.cross (-Eigen::Vector3f (plane.segment (0, 3))).dot (vc1);
+    const float cos_val = vc0.dot (vc1);
+    float angle = std::atan2 (sin_val, cos_val);
+    if (angle < 0.0f) // -pi~pi -> 0~2pi
+    {
+      angle += (float) (2.0 * M_PI);
+    }
+    return angle;
+  }
+  
+  /**
+   * estimate a radius for the ball pivoting algorithm. the returned estimation is the minimal value, so that
+   * at least min_success_rate of the sample points have at least num_in_radius neighbors
+   * @tparam PointNT
+   * @param kdtree
+   * @param num_sample_point
+   * @param num_in_radius
+   * @param min_success_rate
+   * @return
+   */
+  template<typename PointNT>
+  double
+  guess_radius (const typename PCLSurfaceBase<PointNT>::KdTreePtr &kdtree,
```

Comment:
I do not see a reason to have this as a free function. This simply provides an implementation for `setEstimatedRadius()` and is only used there.

BTW, these are the default parameters I asked for.